txnsync/peer.go

```diff
@@ -167,6 +171,82 @@ type Peer struct {
 	lastSelectedTransactionsCount int
 }
 
+// requestParamsGroupCounterState stores the latest group counters for a given set of request params.
+// we use this to ensure we can have multiple iteration of bloom filter scanning over each individual
+// transaction group. This method allow us to reduce the bloom filter errors while avoid scanning the
+// list of transactions redundently.
+//msgp:ignore transactionGroupCounterState
+type requestParamsGroupCounterState struct {
+	offset        byte
+	modulator     byte
+	groupCounters [bloomFilterRetryCount]uint64
+}
+
+// transactionGroupCounterTracker manages the group counter state for each request param.
+//msgp:ignore transactionGroupCounterTracker
+type transactionGroupCounterTracker []requestParamsGroupCounterState
+
+// get returns the group counter for a given set of request param.
+func (t *transactionGroupCounterTracker) get(offset, modulator byte) uint64 {
+	i := t.index(offset, modulator)
+	if i >= 0 {
+		return (*t)[i].groupCounters[0]
+	}
+	return 0
+}
+
+// set updates the group counter for a given set of request param. If no such request
+// param currently exists, it create it.
+func (t *transactionGroupCounterTracker) set(offset, modulator byte, counter uint64) {
+	i := t.index(offset, modulator)
+	if i >= 0 {
+		(*t)[i].groupCounters[0] = counter
+		return
+	}
+	// if it doesn't exists -
+	state := requestParamsGroupCounterState{
+		offset:    offset,
+		modulator: modulator,
+	}
+	state.groupCounters[0] = counter
+
+	if len(*t) == maxTransactionGroupTrackers {
+		// shift all entries by one.
+		copy((*t)[0:], (*t)[1:])
+		(*t)[maxTransactionGroupTrackers-1] = state
+	} else {
+		*t = append(*t, state)
+	}
+}
+
+// roll the counters for a given requests params, so that we would go back and
+// rescan some of the previous transaction groups ( but not all !) when selectPendingTransactions is called.
+func (t *transactionGroupCounterTracker) roll(offset, modulator byte) {
+	i := t.index(offset, modulator)
+	if i < 0 {
+		return
+	}
+
+	if (*t)[i].groupCounters[1] >= (*t)[i].groupCounters[0] {
+		return
+	}
+	firstGroupCounter := (*t)[i].groupCounters[0]
+	copy((*t)[i].groupCounters[0:], (*t)[i].groupCounters[1:])
+	(*t)[i].groupCounters[len((*t)[i].groupCounters)-1] = firstGroupCounter
+}
+
+// index is a helper method for the transactionGroupCounterTracker, helping to locate the index of
+// a requestParamsGroupCounterState in the array that matches the provided request params. The method
+// uses a linear search, which works best against small arrays.
+func (t *transactionGroupCounterTracker) index(offset, modulator byte) int {
+	for i := range *t {
```

Comment:
nit: this loop would be more concise by using the value half of the range iterator, `i, v := range *t`

Reply:
yes, good idea. will do!